airflow/sentry.py

```diff
@@ -0,0 +1,203 @@
+
+        sentry_flask = FlaskIntegration()
+
+        # LoggingIntegration is set by default.
+        integrations = [sentry_flask]
+
+        if executor_name == "CeleryExecutor":
+            from sentry_sdk.integrations.celery import CeleryIntegration
+            from sentry_sdk.integrations.tornado import TornadoIntegration
+
+            sentry_celery = CeleryIntegration()
+            sentry_tornado = TornadoIntegration()
+            integrations += [sentry_celery, sentry_tornado]
+
+        self.task = None
+        self.execution_date = None
+        self.dag_id = None
+
+        dsn = conf.get("sentry", "sentry_dsn")
+        if dsn:
+            init(dsn=dsn, integrations=integrations)
+        else:
+            # Setting up Sentry using environment variables.
+            log.debug("Defaulting to SENTRY_DSN in environment.")
+            init(integrations=integrations)
+
+    def add_tagging(self, task_instance):
+        """
+        Function to add tagging for a task_instance.
+        """
+        self.task = task_instance.task
+        self.execution_date = task_instance.execution_date
+        self.dag_id = task_instance.dag_id
+
+        with configure_scope() as scope:
+            for tag_name in self.SCOPE_TAGS:
+                attribute = getattr(task_instance, tag_name)
+                if tag_name == "operator":
+                    attribute = self.task.__class__.__name__
+                scope.set_tag(tag_name, attribute)
+
+    @provide_session
+    def add_breadcrumbs(self, session=None):
+        """
+        Function to add breadcrumbs inside of a task_instance.
+        """
+
+        if self.task is None or session is None:
+            return
+
+        dag = self.task.dag
+        task_ids = dag.task_ids
+        task_instances = get_task_instances(
+            self.dag_id, task_ids, self.execution_date, session=session
+        )
+
+        for ti in task_instances:
+            if ti is not None and (
+                ti.state == State.SUCCESS or ti.state == State.FAILED
+            ):
+                data = {}
+                for crumb_tag in self.SCOPE_CRUMBS:
+                    data[crumb_tag] = getattr(ti, crumb_tag)
+
+                add_breadcrumb(category="upstream_tasks", data=data, level="info")
+
+    def format_run_task(self, func):
+        """
+        Function for formatting TaskInstance._run_raw_task.
+        """
+
+        @wraps(func)
+        def wrapper(task_instance, *args, session=None, **kwargs):
+            # Wrapping the _run_raw_task function with push_scope to contain
+            # tags and breadcrumbs to a specific Task Instance
+            with push_scope():
+                self.add_tagging(task_instance)
+                self.add_breadcrumbs(session)
+                try:
+                    func(task_instance, *args, session=session, **kwargs)
+                except Exception as e:
+                    capture_exception(e)
```

Comment:
Don't provide `e` here (bare call without arguments) and use a bare `raise` to preserve stacktraces in Python 2

Reply:
We dropped support for Python. 2 on master.